### PR TITLE
Remove CustomElementRegistry

### LINKED ISSUE #2486: Require explicit registration for template-in-template
Automatically initializing Java instances for any eligible tag name encountered in a template is problematic for several reasons:
1. It requires scanning the entire classpath to discover all `@Tag` annotations. This is especially problematic in Spring Boot.
2. It can cause existing code to start malfunctioning when a seemingly unrelated add-on is added to the classpath.
3. It makes it impossible to use specific tags in sub templates even though the custom element implementation is self-sustaining without a server-side Java implementation.
4. There is no way for the application developer to resolve conflicts if there are multiple components with the same `@Tag`, e.g. if one is the subclass of the other.
5. The registry is implemented as a static singleton, which might cause problems with some special classloader setups, e.g. as encountered with Spring Boot DevTools.

All these issues would be resolved if a `PolymerTemplate` would have to explicitly declare which component classes should be considered for automatic instantiation. There is a small risk of discoverability issues with requiring an explicit declaration, but it will hopefully outweigh the drawbacks with automatic discovery.

### PR DESCRIPTION
**Pull request by Legioth**
Instead, `@Uses` should be used to explicitly register classes eligible
for automatic template-in-template injection.

Fixes #2486

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/flow/3598)

### CODE REVIEW

**Comment by Legioth** [comment]
Review status: 0 of 19 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3598#-L61OB5YxmVIqC8_z8Ar:-L61OB5ZIoucaX3Ev9dY:b-6ija2i) (no related file):*
Note that this is a breaking change that also requires documentation updates.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by Legioth** [comment]
Review status: 0 of 19 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3598#-L61OB5YxmVIqC8_z8Ar:-L61SBnP9u88ifqZCSft:b2ba9mu) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Note that this is a breaking change that also requires documentation updates.
</blockquote></details>

In particular, https://github.com/vaadin/flow-and-components-documentation/blob/master/polymer-templates/tutorial-template-subtemplate.asciidoc needs to be revised.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by caalador** [comment]
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3598#-L61OB5YxmVIqC8_z8Ar:-L7YBExY7HxCGSQCtY5i:b-5vv7dt) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

In particular, https://github.com/vaadin/flow-and-components-documentation/blob/master/polymer-templates/tutorial-template-subtemplate.asciidoc needs to be revised.
</blockquote></details>

I for one am for this change as even though the automatic instantiation is nice in it's own way it adds some unneeded wtf moments when thing can just happen.

---

*[flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java, line 304 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3598#-L7Y8aIh9aiiorhk5zUC:-L7Y8aIh9aiiorhk5zUD:b-6xm3hc) ([raw file](https://github.com/vaadin/flow/blob/cd63d3a666041f6ba90bd5dd58b8e7a97094f59e/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L304)):*
> ```Java
>                                         + previous.getName());
>                             }
>                         }));
> ```

This one should be refactored as it is really hard to follow.
Just moving the AnnotationValuesFor to a variable would help a lot and having the forEach as a old type for each `for(Uses usedType: annotationValues)` would lessen the brain strain of the get is present part.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3598)*

**Comment by gilberto-torrezan** [comment]
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java, line 253 at r1](https://reviewable.io:443/reviews/vaadin/flow/3598#-L7UWN8CN39dU-Bxwo8_:-L7UWN8CN39dU-Bxwo8a:bt1gyck) ([raw file](https://github.com/vaadin/flow/blob/cd63d3a666041f6ba90bd5dd58b8e7a97094f59e/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L253)):*
> ```Java
>         Class<?> fieldType = field.getType();
>         if (Component.class.isAssignableFrom(fieldType)) {
>             attachComponentIfUses(element);
> ```

What about performance? How slower or faster this approach is compared to the previous one?

---

*[flow-server/src/main/java/com/vaadin/flow/dom/Element.java, line 118 at r1](https://reviewable.io:443/reviews/vaadin/flow/3598#-L7USsupxgVwRd-BnWR3:-L7USsupxgVwRd-BnWR4:b-6irs0f) ([raw file](https://github.com/vaadin/flow/blob/ce615b04576a41c9b856f189ed361404f6fa2258/flow-server/src/main/java/com/vaadin/flow/dom/Element.java#L118)):*
> ```Java
>      *            autocreate and wire custom element component for element
>      */
>     public Element(String tag, boolean autocreate) {
> ```

Breaking change warning: there are some components that use this constructor internally, such as Notification and Dialog (probably there are more).

It would be better to remove this API before final, but the transition would be smoother if this constructor would be put as `@Deprecated` for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by Legioth** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java, line 253 at r1](https://reviewable.io:443/reviews/vaadin/flow/3598#-L7UWN8CN39dU-Bxwo8_:-L7Z9s1Az-pTV60x2srw:b-8qy80q) ([raw file](https://github.com/vaadin/flow/blob/cd63d3a666041f6ba90bd5dd58b8e7a97094f59e/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L253)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

What about performance? How slower or faster this approach is compared to the previous one?
</blockquote></details>

I don't see any reason for any significant difference. Both the new and the old implementations do O(1) atomic reads and O(1) map lookups per element that is checked.

---

*[flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java, line 304 at r1](https://reviewable.io:443/reviews/vaadin/flow/3598#-L7Y8aIh9aiiorhk5zUC:-L7ZBQP_0KXyT6pXo74u:b-frgms6) ([raw file](https://github.com/vaadin/flow/blob/cd63d3a666041f6ba90bd5dd58b8e7a97094f59e/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L304)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

This one should be refactored as it is really hard to follow.
Just moving the AnnotationValuesFor to a variable would help a lot and having the forEach as a old type for each `for(Uses usedType: annotationValues)` would lessen the brain strain of the get is present part.
</blockquote></details>

Wouldn't want to dump the stream into a list or array just to be able to use an old fashioned `for` loop.

I tried to split the logic up in a different way instead.

---

*[flow-server/src/main/java/com/vaadin/flow/dom/Element.java, line 118 at r1](https://reviewable.io:443/reviews/vaadin/flow/3598#-L7USsupxgVwRd-BnWR3:-L7ZIJIM398kC_srfk5T:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/ce615b04576a41c9b856f189ed361404f6fa2258/flow-server/src/main/java/com/vaadin/flow/dom/Element.java#L118)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Breaking change warning: there are some components that use this constructor internally, such as Notification and Dialog (probably there are more).

It would be better to remove this API before final, but the transition would be smoother if this constructor would be put as `@Deprecated` for now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by caalador** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by Legioth** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3598#-L61OB5YxmVIqC8_z8Ar:-L7ZNZHakPJrAL1jH8wF:bdjkhss) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

I for one am for this change as even though the automatic instantiation is nice in it's own way it adds some unneeded wtf moments when thing can just happen.
</blockquote></details>

Updated documentation in https://github.com/vaadin/flow-and-components-documentation/pull/39

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by gilberto-torrezan** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3598)*

**Comment by caalador** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3598#-:-L9ETElGEkXDqYc9S4HG:bnfp4nl)*
